File: bakong_payment/generate_payment_token.js
const axios = require('axios');
// const qrcode = require('qrcode-terminal');

/**
 * BAKONG TERMINAL PAYMENT GENERATOR
 * This script requests a KHQR from the Bakong API and renders it in the terminal.
 */
async function createAndShowBakongQR(amount, currency = "USD") {
    // API Configuration
    const API_URL = "https://api-bakong.nbc.org.kh/v1/generate_bakong_checkout";
    const API_TOKEN = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJkYXRhIjp7ImlkIjoiZDJmNmMzOGYzNDc2NGM3MSJ9LCJpYXQiOjE3Njk0MTc3MTYsImV4cCI6MTc3NzE5MzcxNn0.Z_w1fHx74NVcVey9VQ3mtaZxeduWC-zYWLj07y33ncI"; 
    const MERCHANT_ID = "choeng_rayu@aclb";

    const payload = {
        merchantId: MERCHANT_ID,
        amount: amount,
        currency: currency,
        description: "Terminal Payment Test",
        externalTransactionId: `TXN_${Date.now()}` // Unique ID per request
    };

    try {
        console.log("\n========================================");
        console.log("   GENERATING BAKONG KHQR PAYMENT");
        console.log("========================================\n");
        
        const response = await axios.post(API_URL, payload, {
            headers: { 
                'Authorization': `Bearer ${API_TOKEN}`,
                'Content-Type': 'application/json'
            }
        });

        // Bakong API response usually nests the result in a 'data' object
        const result = response.data.data || response.data;

        if (result && result.qrCode) {
            const { qrCode, md5 } = result;

            console.log(`💰 Amount : ${amount} ${currency}`);
            console.log(`🆔 Ref MD5: ${md5}`);
            console.log("\n👇 SCAN TO PAY WITH BAKONG APP:\n");
            console.log(qrCode);

            // Renders the KHQR string as a scannable graphic in terminal
            // qrcode.generate(qrCode, { small: true }); 

            console.log("\n[Status: Awaiting Scan...]");
        } else {
            console.log("❌ Failed to parse QR data from Bakong.");
            console.log("Response:", response.data);
        }

    } catch (error) {
        console.error("❌ ERROR CONNECTING TO BAKONG:");
        if (error.response) {
            // Server responded with an error status (4xx, 5xx)
            console.error(`Status: ${error.response.status}`);
            console.error(JSON.stringify(error.response.data, null, 2));
        } else {
            // Network or setup error
            console.error(error.message);
        }
    }
}

// Run the function
createAndShowBakongQR(1.00, "USD");

File: bakong_payment/generate_payment_token.py
import requests
import time
import sys

def create_and_show_bakong_qr(amount, currency="USD", use_mock=False):
    """
    Generate a Bakong payment token and display only the token in terminal.
    """
    # API Configuration
    API_URL = "https://api-bakong.nbc.org.kh/v1/generate_bakong_checkout"
    API_TOKEN = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJkYXRhIjp7ImlkIjoiZDJmNmMzOGYzNDc2NGM3MSJ9LCJpYXQiOjE3Njk0MTc3MTYsImV4cCI6MTc3NzE5MzcxNn0.Z_w1fHx74NVcVey9VQ3mtaZxeduWC-zYWLj07y33ncI"
    MERCHANT_ID = "choeng_rayu@aclb"
    
    # Generate unique transaction ID
    payload = {
        "merchantId": MERCHANT_ID,
        "amount": amount,
        "currency": currency,
        "description": "Terminal Payment Test",
        "externalTransactionId": f"TXN_{int(time.time() * 1000)}"
    }

    try:
        print("\n========================================")
        print("   GENERATING BAKONG KHQR PAYMENT")
        print("========================================\n")
        
        # Mock mode for testing when API is unavailable
        if use_mock:
            print("⚠️  USING MOCK DATA (API unreachable)")
            result = {
                'qrCode': f'00020126360012kh.bnb.khqr01051234567890210201012300398000300059850011480011KHQR0000102050019606400000000100000000000061{int(amount * 100)}5802KH5910MERCHANT6007Phnom Penh63046B3C',
                'md5': 'mock_md5_hash_' + str(int(time.time()))
            }
            qr_code = result.get('qrCode')
            md5_hash = result.get('md5')

            print(f"💰 Amount : {amount} {currency}")
            print(f"🆔 Ref MD5: {md5_hash}")
            print("\n👇 PAYMENT TOKEN:\n")
            print(qr_code)
            print("\n[Status: Mock Mode - Awaiting Scan...]")
            return qr_code
        
        response = requests.post(
            API_URL,
            json=payload,
            headers={
                'Authorization': f'Bearer {API_TOKEN}',
                'Content-Type': 'application/json'
            },
            timeout=10
        )

        # Bakong API response usually nests the result in a 'data' object
        response_data = response.json()
        result = response_data.get('data') or response_data

        if response.status_code == 200 and result and result.get('qrCode'):
            qr_code = result.get('qrCode')
            md5_hash = result.get('md5')

            print(f"💰 Amount : {amount} {currency}")
            print(f"🆔 Ref MD5: {md5_hash}")
            print("\n👇 PAYMENT TOKEN:\n")
            print(qr_code)
            print("\n[Status: Awaiting Scan...]")
            
            return qr_code
        else:
            print("❌ Failed to parse token data from Bakong.")
            print(f"Status: {response.status_code}")
            print(response.text)
            return None

    except Exception as error:
        print("❌ ERROR CONNECTING TO BAKONG:")
        print(f"   {str(error)}")
        print("\n💡 HINT: Use --mock flag to test without API:")
        print("   python3 generate_payment_token.py --mock")
        return None

# Run the script
if __name__ == "__main__":
    use_mock = "--mock" in sys.argv
    create_and_show_bakong_qr(amount=1.00, currency="USD", use_mock=use_mock)